src/passes/MemoryPacking.cpp

```diff
@@ -23,151 +38,607 @@
+  Replacements& replacements,
+  const Index segmentIndex) {
+  // If there was no transformation, only update the indices
+  if (ranges.size() == 1 && !ranges.front().isZero) {
+    for (auto referrer : referrers) {
+      replacements[referrer] = [referrer, segmentIndex](Function*) {
+        if (auto* init = referrer->dynCast<MemoryInit>()) {
+          init->segment = segmentIndex;
+        } else if (auto* drop = referrer->dynCast<DataDrop>()) {
+          drop->segment = segmentIndex;
+        } else {
+          WASM_UNREACHABLE("Unexpected bulk memory operation");
+        }
+        return referrer;
+      };
+    }
+    return;
+  }
 
-      void visitMemoryInit(MemoryInit* curr) {
-        if (!getModule()->memory.segments[curr->segment].isPassive) {
-          Builder builder(*getModule());
-          // trap if (dest > memory.size | offset | size) != 0
-          replaceCurrent(builder.makeIf(
-            builder.makeBinary(
-              OrInt32,
-              builder.makeBinary(
-                GtUInt32,
-                curr->dest,
-                builder.makeBinary(
-                  MulInt32,
-                  builder.makeConst(Literal(Memory::kPageSize)),
-                  builder.makeHost(MemorySize, Name(), {}))),
-              builder.makeBinary(OrInt32, curr->offset, curr->size)),
-            builder.makeUnreachable()));
-          changed = true;
+  Builder builder(*module);
+
+  Name dropStateGlobal;
+
+  // Return the drop state global, initializing it if it does not exist. This
+  // may change module-global state and has the important side effect of setting
+  // dropStateGlobal, so it must be evaluated eagerly, not in the replacements.
+  auto getDropStateGlobal = [&]() {
+    if (dropStateGlobal != Name()) {
+      return dropStateGlobal;
+    }
+    dropStateGlobal = Name(std::string("__mem_segment_drop_state_") +
+                           std::to_string(dropStateGlobalCount++));
+    module->addGlobal(builder.makeGlobal(dropStateGlobal,
+                                         Type::i32,
+                                         builder.makeConst(Literal(int32_t(0))),
+                                         Builder::Mutable));
+    return dropStateGlobal;
+  };
+
+  // Create replacements for memory.init instructions first
+  size_t initIndex = segmentIndex;
+  for (auto referrer : referrers) {
+    auto* init = referrer->dynCast<MemoryInit>();
+    if (init == nullptr) {
+      continue;
+    }
+
+    // Nonconstant offsets or sizes will have inhibited splitting
+    size_t start = init->offset->cast<Const>()->value.geti32();
+    size_t end = start + init->size->cast<Const>()->value.geti32();
+
+    // Index of the range from which this memory.init starts reading
+    size_t firstRangeIdx = 0;
+    while (firstRangeIdx < ranges.size() &&
+           ranges[firstRangeIdx].end <= start) {
+      ++firstRangeIdx;
+    }
+
+    // Handle zero-length memory.inits separately
+    if (start == end) {
```

Comment:
When the length is zero, the `start` may be one past the end of the segment and `firstRangeIdx` may be equal to `ranges.size()`. We handle this case separately so we can assume the start is in bounds and that some range will be intersected, which simplifies subsequent logic. I've updated the comment to reflect this rationale.

Reply:
If `start` is past the end of the segment, shouldn't we have already removed that in `MemoryPacking::optimizeBulkMemoryOps`?

Reply:
If `start` is more than one past the end, it will have been optimized. But if `start` is *at* the end (i.e. one past the end) and the size is zero, then whether or not it will trap depends on the drop state, so it cannot be optimized in `optimizeBulkMemoryOps`.